### DISCUSSION TITLE
pipsolar ESPHome compatibility with AXPERT MAX 11KW TWIN

### DISCUSSION THREAD

**Discussion by jopl** [syssi/esphome-pipsolar#135] [Q&A] [answered]
Dear esphome-pipsolar enthusiastics,
as I have got the AXPERT MAX 11KW TWIN inverter and connect it with ESP-Home successfully, I can read from the logs the right CRC for "QMODI\xC1\r" and "QFLAG\x98t\r" messages. Unfortunately the "QPIRI\xF8T\r" and "QPIGS\xB7\xA9\r" (but not only) are finished by CRC NOK. I tried to identify the issue and found out the protocol is evidently different, i.e.:

>>> "QPIRI\xF8T\r"
[22:08:15][D][pipsolar:837]: Sending polling command : QPIRI with length 5
[22:08:15][D][uart_debug:158]: <<< "(230.0 47.8 230.0 50.0 47.8 11000 11000 48.0 50.0 44.0 55.2 55.2 2 002 100 1 2 3 9 01 0 0 54.0 0 1 480 0 000\xBA\xF6\r"
[22:08:15][D][pipsolar:773]: checking crc on incoming message
[22:08:15][D][pipsolar:782]: CRC NOK expected: 1C 66 but got: 0 0
[22:08:20][D][uart_debug:158]: >>> "QPIGS\xB7\xA9\r"
[22:08:20][D][pipsolar:837]: Sending polling command : QPIGS with length 5
[22:08:20][D][uart_debug:158]: <<< "(245.2 50.1 229.9 50.0 00551 00435 005 368 52.40 000 082 0034 00.0 000.0 00.00 00008 00010000 00 00 00000 010 0 00 0000T\x16\r"
[22:08:20][D][pipsolar:773]: checking crc on incoming message
[22:08:20][D][pipsolar:782]: CRC NOK expected: E2 E3 but got: 54 16

Is there any solution for setting different protocol than actual, please. And how could be set this?
Or where I can change the parsing the incoming messages to be able to read at least some information, please.
Thank you so much for reply in advance.
Kind regards, Josef.

**Comment by jopl** [comment, marked as answer]
... who is interested: in the meanwhile, I found out addition (uncommenting) of
```
external_components:
  - source: github://syssi/esphome-pipsolar@2424mse1
    refresh: 0s
```
plus addition of "select" entity from example of @2424mse1 in my yaml, i.e.:
``` 
select:
  - platform: pipsolar
    pipsolar_id: ${inv_id}
    output_source_priority:
      id: ${inv_id}_output_source_priority_select
      name: ${inv_id}_output_source_priority_select
      optionsmap:
        "Utility first": "POP00"
        "Solar only": "POP01"
        "Solar Battery Utility": "POP02"
      statusmap:
        "0": "Utility first"
        "1": "Solar only"
        "2": "Solar Battery Utility"

  - platform: pipsolar
    pipsolar_id: ${inv_id}
    charging_discharging_control:
      # See MAX_Communication_Protocol_20200526.pdf page 18
      name: "${devicename} charging discharging control"
      optionsmap:
        "111": "PBATCD111"
        "011": "PBATCD011"
        "101": "PBATCD101"
        "110": "PBATCD110"
        "010": "PBATCD010"
        "100": "PBATCD100"
        "001": "PBATCD001"
        "000": "PBATCD000"
      statusmap:
        "111": "111"
        "011": "011"
        "101": "101"
        "110": "110"
        "010": "010"
        "100": "100"
        "001": "001"
        "000": "000"
```
helped me and now I am able to read data without CRC NOK :-).
${inv_id} & ${devicename} should be appropriately defined (in substitutions:).
Regards, Josef.

**Reply by zipp00** [reply]
Can you share the entire source code?
